lib/storage/authgateway.go

```diff
@@ -0,0 +1,561 @@
+// respective certificates.
+func (gw *AuthGatewayV2) PrincipalsChanged(other AuthGateway) bool {
+	if principalsChanged(gw.GetSSHPublicAddrs(), other.GetSSHPublicAddrs()) {
+		return true
+	}
+	if principalsChanged(gw.GetKubernetesPublicAddrs(), other.GetKubernetesPublicAddrs()) {
+		return true
+	}
+	if principalsChanged(gw.GetWebPublicAddrs(), other.GetWebPublicAddrs()) {
+		return true
+	}
+	return false
+}
+
+func principalsChanged(old, new []string) bool {
+	oldSet := utils.NewStringSetFromSlice(utils.Hosts(old))
+	newSet := utils.NewStringSetFromSlice(utils.Hosts(new))
+	return len(oldSet.Diff(newSet)) > 0
+}
+
+// SettingsChanged returns true if connection settings are different between
+// this and provided auth gateway configuration.
+func (gw *AuthGatewayV2) SettingsChanged(other AuthGateway) bool {
+	if gw.GetMaxConnections() != other.GetMaxConnections() {
+		return true
+	}
+	if gw.GetMaxUsers() != other.GetMaxUsers() {
+		return true
+	}
+	if gw.GetClientIdleTimeout() != nil && other.GetClientIdleTimeout() == nil ||
+		gw.GetClientIdleTimeout() == nil && other.GetClientIdleTimeout() != nil ||
+		gw.GetClientIdleTimeout() != nil && other.GetClientIdleTimeout() != nil &&
+			gw.GetClientIdleTimeout().Value() != other.GetClientIdleTimeout().Value() {
+		return true
+	}
+	if gw.GetDisconnectExpiredCert() != nil && other.GetDisconnectExpiredCert() == nil ||
+		gw.GetDisconnectExpiredCert() == nil && other.GetDisconnectExpiredCert() != nil ||
+		gw.GetDisconnectExpiredCert() != nil && other.GetDisconnectExpiredCert() != nil &&
+			gw.GetDisconnectExpiredCert().Value() != other.GetDisconnectExpiredCert().Value() {
+		return true
+	}
+	return false
+}
+
+// ApplyTo applies auth gateway settings to the provided other auth gateway.
+//
+// Only non-nil settings are applied.
+func (gw *AuthGatewayV2) ApplyTo(other AuthGateway) {
+	if v := gw.GetConnectionLimits(); v != nil {
+		other.SetConnectionLimits(*v)
+	}
+	if v := gw.GetClientIdleTimeout(); v != nil {
+		other.SetClientIdleTimeout(*v)
+	}
+	if v := gw.GetDisconnectExpiredCert(); v != nil {
+		other.SetDisconnectExpiredCert(*v)
+	}
+	if v := gw.GetAuthPreference(); v != nil {
+		other.SetAuthPreference(v)
+	}
+	if gw.Spec.PublicAddr != nil {
+		other.SetSSHPublicAddrs(*gw.Spec.PublicAddr)
+		other.SetKubernetesPublicAddrs(*gw.Spec.PublicAddr)
+		other.SetWebPublicAddrs(*gw.Spec.PublicAddr)
+	}
+	if gw.Spec.SSHPublicAddr != nil {
+		other.SetSSHPublicAddrs(*gw.Spec.SSHPublicAddr)
+	}
+	if gw.Spec.KubernetesPublicAddr != nil {
+		other.SetKubernetesPublicAddrs(*gw.Spec.KubernetesPublicAddr)
+	}
+	if gw.Spec.WebPublicAddr != nil {
+		other.SetWebPublicAddrs(*gw.Spec.WebPublicAddr)
+	}
+}
+
+// Apply applies auth gateway settings to the provided config.
+func (gw *AuthGatewayV2) ApplyToTeleportConfig(config *teleconfig.FileConfig) {
+	if gw.Spec.ConnectionLimits != nil {
+		config.Global.Limits.MaxConnections = *gw.Spec.ConnectionLimits.MaxConnections
```

Comment:
`ConnectionLimits.MaxConnections` might still be `nil` (ditto for `MaxUsers`).
I'd just use `GetMaxConnections` / `GetMaxUsers`